Mathlib/Data/Tree.lean

```diff
@@ -0,0 +1,180 @@
+/-
+Copyright (c) 2019 mathlib community. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Wojciech Nawrocki
+
+! This file was ported from Lean 3 source module data.tree
+! leanprover-community/mathlib commit ed989ff568099019c6533a4d94b27d852a5710d8
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Lean.Data.RBTree
+import Mathlib.Data.Num.Basic
+import Mathlib.Order.Basic
+import Mathlib.Init.Data.Ordering.Basic
+
+/-!
+# Binary tree
+
+Provides binary tree storage for values of any type, with O(lg n) retrieval.
+See also `Lean.Data.RBTree` for red-black trees - this version allows more operations
+to be defined and is better suited for in-kernel computation.
+
+We also specialize for `Tree Unit`, which is a binary tree without any
+additional data. We provide the notation `a △ b` for making a `Tree Unit` with children
+`a` and `b`.
+
+## References
+
+<https://leanprover-community.github.io/archive/stream/113488-general/topic/tactic.20question.html>
+-/
+
+
+/-- A binary tree with values stored in non-leaf nodes. -/
+inductive Tree.{u} (α : Type u) : Type u
+  | nil : Tree α
+  | node : α → Tree α → Tree α → Tree α
+  deriving DecidableEq
+-- Porting note: Removed `deriving has_reflect`.
+#align tree Tree
+
+namespace Tree
+
+universe u
+
+variable {α : Type u}
+
+/-- Construct a string representation of a tree. Provides a `hasRepr` instance. -/
+def repr [Repr α] : Tree α → String
+  | nil => "nil"
+  | node a t1 t2 => "Tree.node " ++ reprStr a ++ " (" ++ repr t1 ++ ") (" ++ repr t2 ++ ")"
+#align tree.repr Tree.repr
+
+instance [Repr α] : Repr (Tree α) :=
+  ⟨fun s _ => Tree.repr s⟩
```

Comment:
`Repr` in Lean4 expects you to produce a `Std.Format`, not a `String`. I think it would make sense to change `repr` accordingly. @digama0: is it better to `#noalign`, or use the subscript `x`?

Reply:
It's also possible that `deriving Repr` will work.

Reply:
Done